File: leetcode/226-invert-binary-tree/invert_binary_tree_bfs_dfs.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Date    : 2024/08/13 14:48:59
# @Author  : Kelley HUANG (dianhuangkan@gmail.com)
# @Link    : https://leetcode.com/problems/invert-binary-tree/

# 递归、BFS、DFS，3种遍历方式实现翻转
from typing import Optional

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def invertTreeBFS(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        """invertTreeBFS
        层序遍历方式翻转
        """
        # 用队列实现
        if root == None:
            return root
        queue = []
        queue.append(root)
        while not len(queue) == 0:
            node = queue.pop(0)
            node.left, node.right = node.right, node.left
            if node.left != None:
                queue.append(node.left)
            if node.right != None:
                queue.append(node.right)

        return root


class Solution:
    def invertTreeDFS(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        """深度优先遍历的方式翻转
        """
        # 用栈实现
        if root == None:
            return root
        queue = []
        queue.append(root)
        while not len(queue) == 0:
            node = queue.pop()
            node.left, node.right = node.right, node.left
            # 先进后出，则left子节点先被访问
            if node.right != None:
                queue.append(node.right)
            if node.left != None:
                queue.append(node.left)

        return root

File: leetcode/226-invert-binary-tree/invert_binary_tree_bfs_dfs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */

#include <queue>
#include <stack>
using std::queue;
using std::stack;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  TreeNode *invertTreeBFS(TreeNode *root) {
    // invertTreeBFS
    // 层序遍历方式翻转
    // 用队列实现
    if (root == nullptr) return root;
    queue<TreeNode *> q;
    q.push(root);
    TreeNode *node;
    while (!q.empty()) {
      node = q.front();
      q.pop();
      TreeNode *tmp = node->left;
      node->left = node->right;
      node->right = tmp;
      if (node->left != nullptr) q.push(node->left);
      if (node->right != nullptr) q.push(node->right);
    }
    return root;
  }

  TreeNode *invertTreeDFS(TreeNode *root) {
    // invertTreeDFS
    // 深度优先遍历的方式翻转
    // 用栈实现
    if (root == nullptr) return root;
    stack<TreeNode *> s;
    s.push(root);
    TreeNode *node;
    while (!s.empty()) {
      node = s.top();
      s.pop();
      TreeNode *tmp = node->left;
      node->left = node->right;
      node->right = tmp;
      if (node->right != nullptr) s.push(node->right);
      if (node->left != nullptr) s.push(node->left);
    }
    return root;
  }
};